gguf-py/gguf/gguf_manager.py

```diff
@@ -0,0 +1,254 @@
+        if not self.shards:
+            self.shards.append(["", 0, METADATA_ONLY_INDICATOR, None])
+
+        # format shard names
+        if len(self.shards) == 1:
+            self.shards[0][0] = self.path
+        else:
+            for i in range(len(self.shards)):
+                self.shards[i][0] = self.path.with_name(SHARD_NAME_FORMAT.format(self.path.stem, i + 1, len(self.shards)))
+
+        # print shard info
+        print("\nWriting the following files:")
+        for (path, tensor_ct, size, _) in self.shards:
+            print(f"  {path}: n_tensors = {tensor_ct}, total_size = {GGUFManager.format_n_bytes_to_str(size)}")
+        print()
+
+        if self.split_arguments.dry_run:
+            print("\nDry run, not writing files")
+            exit()
+
+        # we don't want to initialize GGUFWriters until now because they create files
+        for i, (path, _, _, tensors) in enumerate(self.shards):
+            # dont_add_architecture is used for consistency - examples/gguf_split doesn't add arch to all shards
+            writer = GGUFWriter(path, self.arch, use_temp_file=self.use_temp_file,
+                                endianess=self.endianess, dont_add_architecture=not (i == 0))
+
+            # only the first shard needs all the KV data
+            if i == 0:
+                for key, (value, etype) in self.kv_data.items():
+                    writer.add_key(key)
+                    writer.add_val(value, etype)
+
+            # add split metadata unless it's one file - small first shard splits even with SplitStyle.NONE
+            if self.split_arguments.split_style != SplitStyle.NONE or self.split_arguments.small_first_shard:
+                writer.add_uint16(LLM_KV_SPLIT_NO, i)
+                writer.add_uint16(LLM_KV_SPLIT_COUNT, len(self.shards))
+                writer.add_int32(LLM_KV_SPLIT_TENSORS_COUNT, self.total_tensors)
+
+            # add tensors, deque popleft() ensures references to eager tensors are not kept
+            while True:
+                try:
+                    (name, tensor, dtype) = tensors.popleft()
+                    writer.add_tensor(name, tensor, raw_dtype=dtype)
+                except:
+                    break
+
+            self.shard_writers.append(writer)
+
+    def write_header_to_file(self) -> None:
+        if self.state is not WriterState.EMPTY:
+            raise ValueError(f'Expected GGUFManager state to be EMPTY, got {self.state}')
+
+        for writer in self.shard_writers:
+            writer.write_header_to_file()
+
+        self.state = WriterState.HEADER
+
+    def write_kv_data_to_file(self) -> None:
+        if self.state is not WriterState.HEADER:
+            raise ValueError(f'Expected GGUFManager state to be HEADER, got {self.state}')
+        
+        for writer in self.shard_writers:
+            writer.write_kv_data_to_file()
+
+        self.state = WriterState.KV_DATA
+
+    def write_tensors_to_file(self, progress: bool = False) -> None:
```

Comment:
```suggestion
    def write_tensors_to_file(self, *, progress: bool = False) -> None:
```

This is otherwise incompatible with `GGUFWriter`. It's only accepted as a keyword argument, not also positional.